python/ray/dag/compiled_dag_node.py

```diff
@@ -334,34 +366,60 @@ def _get_or_compile(
             elif isinstance(task.dag_node, InputNode):
                 task.output_channel = Channel(
                     buffer_size_bytes=self._buffer_size_bytes,
-                    num_readers=task.num_readers,
+                    num_readers=task.num_readers if task.num_readers else 1,
                 )
             else:
-                assert isinstance(task.dag_node, MultiOutputNode)
+                assert isinstance(task.dag_node, MultiOutputNode) or isinstance(
+                    task.dag_node, InputAttributeNode
+                )
 
             for idx in task.downstream_node_idxs:
                 queue.append(idx)
 
         for node_idx, task in self.idx_to_task.items():
-            if node_idx == self.input_task_idx:
-                # We don't need to assign an actual task for the input node.
+            if node_idx in self.input_task_idxs:
+                # We don't need to assign an actual task for the input node(s).
                 continue
 
             if node_idx == self.output_task_idx:
-                # We don't need to assign an actual task for the input node.
+                # We don't need to assign an actual task for the output node.
                 continue
 
             resolved_args = []
+            # for single arg, read_idx is -1, meaning taking the whole input,
+            # for multiple args, read_idx means the index of element to take,
+            # Example:
+            # inp = InputNode()
+            # f.remote(inp.x)
+            # g.remote(inp.y)
+            # f and g both read (x, y) from the same channel, but pick different indices
+            read_idxs = []
             has_at_least_one_channel_input = False
             for arg in task.args:
                 if isinstance(arg, DAGNode):
-                    arg_idx = self.dag_node_to_idx[arg]
-                    arg_channel = self.idx_to_task[arg_idx].output_channel
+                    parent_node = (
+                        arg._dag_input_node
+                        if isinstance(arg, InputAttributeNode)
+                        else arg
+                    )
+                    parent_idx = self.dag_node_to_idx[parent_node]
+                    arg_channel = self.idx_to_task[parent_idx].output_channel
                     assert arg_channel is not None
                     resolved_args.append(arg_channel)
+                    read_idx = -1
+                    if isinstance(arg, InputAttributeNode):
+                        # use arg idx - 1 here because when there are multiple args,
+                        # they are ordered in InputNode, InputAttributeNode,
+                        # InputAttributeNode, ..., while only InputAttributeNodes
+                        # carry arguments
+                        read_idx = self.dag_node_to_idx[arg] - 1
```

Comment:
Along lines of above comment, if we collapse all InputAttributeNode/InputNodes into one DAG node, this can be simpler and more robust to changes in the DAG node_to_idx ordering.

Reply:
Agreed, I can make a followup pr to pack InputAttributeNode/InputNodes into one DAG node, feel like that can go beyond the scope of this pr